<img src="images/laliga.png">

# La_Liga_2018_19

* **Season 2018-19:**
   
   The *2018–19 La Liga* season, also known as *La Liga Santander* for sponsorship reasons, was the 88th since its establishment. The season began on 17 August 2018 and finished on 26 May 2019. Fixtures for the 2018–19 season were announced on 24 July 2018. This was the *first La Liga season using VAR*. 
   
   *Barcelona* were the *defending champions* after winning La Liga last season for the *25th time*. *Huesca, Rayo Vallecano and Valladolid* join as the promoted clubs from the *2017–18 Segunda División*. They replaced *Málaga, Las Palmas and Deportivo La Coruña* who were relegated to the *2018–19 Segunda División*. *Barcelona* were crowned as *league winners* for the second season in a row on the 28 April, after defeating *Levante 1–0 with 3 games to spare*.
   
   
* **About the Project:**

    In this project we will try to analyze the data from La Liga 2018-19 season. We will try to analyze the pattern in the data which will help us to gain more insight about the La Liga season.

### 1. The Imports:

* **Packages Imported** :-

    1. sys: access to system parameters.
    2. numpy : for numerical computation.
    3. pandas: for data manipulation and analysis.
    4. matplotlib: plotting library.    
    5. seaborn: plotting library based on matplotlib.

In [25]:
## load packages
import sys
import numpy as np
import pandas as pd
import matplotlib as mtlb
import seaborn as sns

In [24]:
## version of packages used
print(f'Python version: {sys.version}', end='\n\n')
print(f'Numpy version: {np.__version__}')
print(f'Pandas version: {pd.__version__}')
print(f'Matplotlib version: {mtlb.__version__}')
print(f'Seaborn version: {sns.__version__}')

Python version: 3.6.8 |Anaconda custom (64-bit)| (default, Feb 21 2019, 18:30:04) [MSC v.1916 64 bit (AMD64)]

Numpy version: 1.12.0
Pandas version: 0.23.4
Matplotlib version: 3.1.0
Seaborn version: 0.9.0


In [3]:
## set pandas to show all columns
pd.set_option('display.max_columns', None)

## import style and pyplot from matplotlib
import matplotlib.style as style
import matplotlib.pyplot as plt

## set style for our plot
style.use('ggplot')

## inline magic function
%matplotlib inline

### 2. The Data:

* **Overview:**

  The dataset contains information about each and every match played in the La Liga season 2018-19, and with the help of the data we will try to gain insight about the season.
  
  The dataset is downloaded from: http://football-data.co.uk/data.php

* The dataset contains following fields:

    * Div : League Division
    * Date : Match Date (dd/mm/yy)
    * HomeTeam : Home Team
    * AwayTeam : Away Team
    * FTHG : Full Time Home Team Goals
    * FTAG : Full Time Away Team Goals
    * FTR : Full Time Result (H=Home Win, D=Draw, A=Away Win)
    * HTHG : Half Time Home Team Goals
    * HTAG : Half Time Away Team Goals
    * HTR : Half Time Result (H=Home Win, D=Draw, A=Away Win)
    * HS : Home Team Shots
    * AS : Away Team Shots
    * HST : Home Team Shots on Target
    * AST : Away Team Shots on Target
    * HF : Home Team Fouls Committed
    * AF : Away Tema Fouls Committed
    * HC : Home Team Corners
    * AC : Away Team Corners
    * HY : Home Team Yellow Cards
    * AY : Away Team Yellow Cards
    * HR : Home Team Red Cards
    * AR : Away Team Red Cards
    

* Note: We are not including the *betting columns* in our project.

* Read the data from laliga_2018_19.csv file.

In [4]:
## laliga will store the dataset from laliga_2018_19.csv
laliga = pd.read_csv('laliga_2018_19.csv')

* Use the basic functions to see the dataset and the information about the dataset.

In [5]:
## head() to see first 10 rows of the dataset
laliga.head(10)

Div        Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  SP1  17/08/2018        Betis     Levante     0     3   A     0     1   A   
1  SP1  17/08/2018       Girona  Valladolid     0     0   D     0     0   D   
2  SP1  18/08/2018    Barcelona      Alaves     3     0   H     0     0   D   
3  SP1  18/08/2018        Celta     Espanol     1     1   D     0     1   A   
4  SP1  18/08/2018   Villarreal    Sociedad     1     2   A     1     1   D   
5  SP1  19/08/2018        Eibar      Huesca     1     2   A     0     2   A   
6  SP1  19/08/2018  Real Madrid      Getafe     2     0   H     1     0   H   
7  SP1  19/08/2018    Vallecano     Sevilla     1     4   A     0     3   A   
8  SP1  20/08/2018   Ath Bilbao     Leganes     2     1   H     1     1   D   
9  SP1  20/08/2018     Valencia  Ath Madrid     1     1   D     0     1   A   

   HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  B365H  B365D  B365A  \
0  22   6    8    4  10  10   5   3   0   2   0   0   1.66   4.00   5.00   
1  13   2    1    1  21  20   3   2   1   1   0   0   1.75   3.60   5.00   
2  25   3    9    0   6  13   7   1   0   2   0   0   1.11  10.00  21.00   
3  12  14    2    5  13  14   8   7   3   2   0   0   1.85   3.50   4.50   
4  16   8    7    4  16  10   4   6   2   3   0   0   2.04   3.40   3.80   
5  18   8    6    6  12  13   7   0   1   1   0   0   1.66   3.75   5.50   
6  10   4    3    1  11  27   3   0   1   7   0   0   1.20   7.00  13.00   
7  13  17    2    8   6  15   2   6   1   0   0   0   3.25   3.60   2.14   
8  17  12    5    2  12  13   6   2   4   5   0   0   1.75   3.30   5.50   
9  13   9    4    3  10  15   4  10   2   3   0   0   3.00   3.20   2.50   

    BWH    BWD    BWA   IWH   IWD    IWA   PSH    PSD    PSA   WHH  WHD  \
0  1.70   3.70   5.25  1.75  3.60   4.90  1.69   4.19   5.11  1.67  3.9   
1  1.75   3.50   5.25  1.80  3.60   4.50  1.80   3.70   4.99  1.75  3.6   
2  1.11  10.00  20.00  1.12  9.00  20.00  1.11  11.27  25.40  1.08  9.0   
3  1.91   3.40   4.25  1.90  3.50   4.10  1.93   3.64   4.27  1.91  3.5   
4  2.05   3.30   3.90  2.00  3.40   3.80  2.06   3.51   3.91  2.05  3.3   
5  1.70   3.70   5.25  1.70  3.75   5.00  1.72   3.90   5.26  1.73  3.6   
6  1.18   7.25  16.00  1.20  6.50  15.00  1.20   7.36  17.47  1.22  6.0   
7  3.50   3.50   2.10  3.50  3.40   2.10  3.46   3.74   2.13  3.30  3.7   
8  1.78   3.50   5.00  1.85  3.50   4.40  1.79   3.54   5.46  1.80  3.4   
9  2.85   3.25   2.55  2.85  3.20   2.55  3.12   3.18   2.57  3.00  3.2   

     WHA   VCH   VCD    VCA  Bb1X2  BbMxH  BbAvH  BbMxD  BbAvD  BbMxA  BbAvA  \
0   4.75  1.67   4.2   5.20     40   1.75   1.68   4.25   4.00   5.25   4.95   
1   4.60  1.80   3.7   4.80     40   1.85   1.78   3.83   3.60   5.27   4.79   
2  29.00  1.10  10.5  34.00     40   1.13   1.10  11.50   9.82  41.00  25.67   
3   4.00  1.93   3.5   4.40     38   1.97   1.90   3.73   3.53   4.50   4.20   
4   3.60  2.05   3.5   3.90     40   2.11   2.03   3.62   3.43   3.93   3.76   
5   4.75  1.70   3.8   5.00     40   1.76   1.70   3.93   3.77   5.50   5.08   
6  13.00  1.20   7.0  13.00     39   1.24   1.21   7.36   6.66  17.47  14.13   
7   2.05  3.40   3.6   2.10     40   3.53   3.38   3.75   3.56   2.20   2.11   
8   4.75  1.80   3.4   5.00     40   1.85   1.78   3.64   3.43   5.50   5.03   
9   2.40  3.00   3.2   2.45     39   3.12   2.99   3.29   3.14   2.61   2.51   

   BbOU  BbMx>2.5  BbAv>2.5  BbMx<2.5  BbAv<2.5  BbAH  BbAHh  BbMxAHH  \
0    38      1.82      1.76      2.15      2.06    20  -0.75     1.89   
1    38      2.21      2.13      1.78      1.71    20  -0.75     2.06   
2    32      1.39      1.34      3.40      3.18    19  -2.50     1.95   
3    36      2.13      2.06      1.84      1.76    18  -0.75     2.26   
4    37      2.05      1.99      1.88      1.81    18  -0.25     1.76   
5    37      1.95      1.88      1.98      1.91    19  -0.75     1.96   
6    33      1.50      1.45      2.75      2.66    19  -1.75     1.85   
7    

* Drop all the *betting columns* from the dataset:
    
    * Since we only want the first 22 columns, all the other columns are basically the betting columns.
    * So we will slice the dataframe which will give us the dataset having first 22 columns.

In [6]:
## slice the dataset
laliga = laliga.iloc[:,:22]

In [7]:
## use head() function to see dataset's first few rows
laliga.head(10)

Div        Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  SP1  17/08/2018        Betis     Levante     0     3   A     0     1   A   
1  SP1  17/08/2018       Girona  Valladolid     0     0   D     0     0   D   
2  SP1  18/08/2018    Barcelona      Alaves     3     0   H     0     0   D   
3  SP1  18/08/2018        Celta     Espanol     1     1   D     0     1   A   
4  SP1  18/08/2018   Villarreal    Sociedad     1     2   A     1     1   D   
5  SP1  19/08/2018        Eibar      Huesca     1     2   A     0     2   A   
6  SP1  19/08/2018  Real Madrid      Getafe     2     0   H     1     0   H   
7  SP1  19/08/2018    Vallecano     Sevilla     1     4   A     0     3   A   
8  SP1  20/08/2018   Ath Bilbao     Leganes     2     1   H     1     1   D   
9  SP1  20/08/2018     Valencia  Ath Madrid     1     1   D     0     1   A   

   HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  
0  22   6    8    4  10  10   5   3   0   2   0   0  
1  13   2    1    1  21  20   3   2   1   1   0   0  
2  25   3    9    0   6  13   7   1   0   2   0   0  
3  12  14    2    5  13  14   8   7   3   2   0   0  
4  16   8    7    4  16  10   4   6   2   3   0   0  
5  18   8    6    6  12  13   7   0   1   1   0   0  
6  10   4    3    1  11  27   3   0   1   7   0   0  
7  13  17    2    8   6  15   2   6   1   0   0   0  
8  17  12    5    2  12  13   6   2   4   5   0   0  
9  13   9    4    3  10  15   4  10   2   3   0   0

In [8]:
## use info() method
laliga.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Data columns (total 22 columns):
Div         380 non-null object
Date        380 non-null object
HomeTeam    380 non-null object
AwayTeam    380 non-null object
FTHG        380 non-null int64
FTAG        380 non-null int64
FTR         380 non-null object
HTHG        380 non-null int64
HTAG        380 non-null int64
HTR         380 non-null object
HS          380 non-null int64
AS          380 non-null int64
HST         380 non-null int64
AST         380 non-null int64
HF          380 non-null int64
AF          380 non-null int64
HC          380 non-null int64
AC          380 non-null int64
HY          380 non-null int64
AY          380 non-null int64
HR          380 non-null int64
AR          380 non-null int64
dtypes: int64(16), object(6)
memory usage: 65.4+ KB


* **Summary of Dataframe -- using info():**

    * There is a total of 380 entries, that means 380 matches were played in the whole season.
    * The dataframe contains 22 columns: 16 have int values and 6 have string values.
    * No column has NaN values.

In [9]:
## use describe method()
laliga.describe()

FTHG        FTAG        HTHG        HTAG          HS          AS  \
count  380.000000  380.000000  380.000000  380.000000  380.000000  380.000000   
mean     1.452632    1.134211    0.544737    0.513158   13.873684   10.434211   
std      1.146461    1.134905    0.681918    0.756562    5.048560    3.952006   
min      0.000000    0.000000    0.000000    0.000000    3.000000    2.000000   
25%      1.000000    0.000000    0.000000    0.000000   10.000000    8.000000   
50%      1.000000    1.000000    0.000000    0.000000   13.000000   10.000000   
75%      2.000000    2.000000    1.000000    1.000000   17.000000   13.000000   
max      8.000000    6.000000    3.000000    5.000000   34.000000   21.000000   

              HST         AST          HF          AF          HC          AC  \
count  380.000000  380.000000  380.000000  380.000000  380.000000  380.000000   
mean     4.834211    3.589474   13.631579   13.450000    5.573684    4.021053   
std      2.484015    2.052065    4.096598    3.992093    2.842820    2.352695   
min      0.000000    0.000000    1.000000    3.000000    0.000000    0.000000   
25%      3.000000    2.000000   11.000000   11.000000    4.000000    2.000000   
50%      5.000000    3.000000   13.000000   13.000000    5.000000    4.000000   
75%      6.000000    5.000000   16.000000   16.000000    7.000000    6.000000   
max     15.000000   11.000000   26.000000   27.000000   15.000000   12.000000   

               HY          AY          HR          AR  
count  380.000000  380.000000  380.000000  380.000000  
mean     2.528947    2.642105    0.086842    0.121053  
std      1.568726    1.413585    0.281975    0.342394  
min      0.000000    0.000000    0.000000    0.000000  
25%      1.000000    2.000000    0.000000    0.000000  
50%      2.000000    3.000000    0.000000    0.000000  
75%      4.000000    3.000000    0.000000    0.000000  
max      8.000000    7.000000    1.000000    2.000000

* **Summary of Dataframe -- using describe():**
    
    * The mean of FTHG(Full Time Home Goals) and FTAG(Full Time Away Goals) comes out to be 1.45 and 1.13 respectively, well that means most matches played in the season ended in a draw. With the full time maximum goals scored by a home team is 8 and by an away team is 6, and the minimum goals scored by both teams is the same which is equal to 0.
    * The mean of HTHG(Half Time Home Goals) and HTAG(Half Time Away Goals) comes out to be 0.54 and 0.51 respectively, same as the above picture, concluding that most matches at half time has their scores draw. With the half time maximum goals scored by a home team is 3 and by an away team is 5, and the minimum goals scored by both teams is the same which is equal to 0.
    * The mean of HS(Home Team Shots) and AS(Away Team Shots) comes out to be 13.87 and 10.43 respectively, with the maximum being 34 and 21 respectively and the minimum being 3 and 2 respectively. At the end the home team has more shots than the away team.
    * The mean of HST(Home Team Shots on Target) and AST(Away Team Shots on Target) comes out to be 4.8 and 3.5 respectively, with maximum being 15 and 11 respectively and minimum being 0 for both teams.
    * The mean of HF(Home Team Foul Committed) and AF(Away Team Foul Committed) comes out to be 13.63 and 13.45 respectively, with maximum being 26 and 27 respectively and minimum being 1 and 3 respectively. At the end the fouls committed by both teams are nearly equal.
    * The mean of HC(Home Team Corners) and AC(Away Team Corners) comes out to be 5.57 and 4.02 respectively, with maximum being 15 and 12 respectively and minimum being 0 for both teams.
    * The mean of HY(Home Team Yellow Cards) and AY(Away Team Yellow Cards) comes out to be 2.52 and 2.64 respectively, with maximum being 8 and 7 respectively and minimum being 0 for both the teams.
    * The mean of HR(Home Team Red Cards) and AR(Away Team Red Cards) comes out to be 0.08 and 0.12 respectively, with maximum being 1 and 2 respectively and minimum being 0 for both the teams.

* Remove *Div* field from the dataset.

In [13]:
## use drop() method
laliga.drop('Div', axis=1, inplace=True)

* Change datatype of Date field.

In [14]:
## original datatype of Date field
type(laliga['Date'].iloc[0])

str

In [27]:
## change the datatype from str to a timestamp object
laliga['Date'] = pd.to_datetime(arg=laliga['Date'], dayfirst=True)
## converting the type of 'Date' field from string value to timestamp object

In [28]:
## changed datatype of Date field
type(laliga['Date'].iloc[0])

pandas._libs.tslibs.timestamps.Timestamp

* Creating two new columns:
    * Second Half Home Goals(SHHG) : goals scored by the home team in the second half. SHHG = FTHG - HTHG
    * Second Half Away Goals(SHAG) : goals scored by the away team in the second half. SHAG = FTAG - HTAG

In [31]:
## create SHHG using FTHG and HTHG
loc = 9
column = 'SHHG'
value = laliga['FTHG'] - laliga['HTHG']
laliga.insert(loc, column, value)

In [32]:
laliga.head(10)

Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  SHHG  \
0 2018-08-17        Betis     Levante     0     3   A     0     1   A     0   
1 2018-08-17       Girona  Valladolid     0     0   D     0     0   D     0   
2 2018-08-18    Barcelona      Alaves     3     0   H     0     0   D     3   
3 2018-08-18        Celta     Espanol     1     1   D     0     1   A     1   
4 2018-08-18   Villarreal    Sociedad     1     2   A     1     1   D     0   
5 2018-08-19        Eibar      Huesca     1     2   A     0     2   A     1   
6 2018-08-19  Real Madrid      Getafe     2     0   H     1     0   H     1   
7 2018-08-19    Vallecano     Sevilla     1     4   A     0     3   A     1   
8 2018-08-20   Ath Bilbao     Leganes     2     1   H     1     1   D     1   
9 2018-08-20     Valencia  Ath Madrid     1     1   D     0     1   A     1   

   HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  
0  22   6    8    4  10  10   5   3   0   2   0   0  
1  13   2    1    1  21  20   3   2   1   1   0   0  
2  25   3    9    0   6  13   7   1   0   2   0   0  
3  12  14    2    5  13  14   8   7   3   2   0   0  
4  16   8    7    4  16  10   4   6   2   3   0   0  
5  18   8    6    6  12  13   7   0   1   1   0   0  
6  10   4    3    1  11  27   3   0   1   7   0   0  
7  13  17    2    8   6  15   2   6   1   0   0   0  
8  17  12    5    2  12  13   6   2   4   5   0   0  
9  13   9    4    3  10  15   4  10   2   3   0   0